src/Bicep.Core/TypeSystem/TypeAssignmentVisitor.cs

```diff
@@ -574,8 +574,15 @@ public override void VisitUnionTypeSyntax(UnionTypeSyntax syntax)
 
                 base.VisitUnionTypeSyntax(syntax);
 
+                if (declaredType is DiscriminatedObjectType or ErrorType
+                    || (declaredType is UnionType unionType && TypeHelper.TryRemoveNullability(unionType) is DiscriminatedObjectType))
```

Comment:
The second clause should be removable per the comment on line 904 of DeclaredTypeManager.

Reply:
https://github.com/Azure/bicep/pull/11222/commits/7046f5c6aea25bee7bea5eb7c685206fb46e5f0b